File: zbs243_AP_FW/packagebinaries.php
<?php

$types[0x00] = "AP_FW_1.54.bin";
$types[0x01] = "AP_FW_2.9.bin";
$types[0xF0] = "AP_FW_Segmented_UK.bin";
$types[0xFF] = "AP_FW_Nodisplay.bin";

$binpath = "../binaries/";
$tocmaxsize = 512;

$toc = array();
$output = '';
$output = str_pad($output,$tocmaxsize,"\0"); 

$version = exec("cat main.c | grep version | grep uint16_t");
list($blaat, $version) = explode("= 0x",$version);
if(strlen($version)<4)die("Couldn't read version from main.c, please check. I got $version");
$version = hexdec($version);

exec("ls -1 $binpath | grep 'AP_FW' | grep -v Pack", $binaries);
foreach($binaries as $file){
	$file = trim($file);
	$type = -1;
	foreach($types as $typeid => $typefile){
		if($typefile == $file){
			$type = $typeid;
		}
	}
	if($type==-1)die("We don't recognize filetype <$file>, sorry...\n");
	$binary = file_get_contents($binpath.$file);
	$length = strlen($binary);
	$offset = strlen($output);
	$subarr['type']=$type;
	$subarr['version']=$version;
	$subarr['name']=$file;
	$subarr['offset']=$offset;
	$subarr['length']=$length;
	$toc[] = $subarr;
	$output.=$binary;
}

$jtoc = json_encode($toc);
$tocsize = strlen($jtoc);
if($tocsize>$tocmaxsize)die("TOC is too big! adjust size and try again\n");
$output = substr_replace($output,$jtoc,0,strlen($jtoc));
file_put_contents($binpath."AP_FW_Pack.bin", $output);
print_r($toc);
echo "All done.\n";

File: zbs243_AP_FW/packagebinaries.py
import os
import json

types = {
    0x00: "AP_FW_1.54.bin",
    0x01: "AP_FW_2.9.bin",
    0xF0: "AP_FW_Segmented_UK.bin",
    0xFF: "AP_FW_Nodisplay.bin"
}

binpath = "../binaries/"
tocmaxsize = 512

toc = []
output = bytearray(tocmaxsize)

# Read version from main.c
with open("main.c") as file:
    lines = file.readlines()
    version_line = next(line for line in lines if "version" in line and "uint16_t" in line)
    _, version = version_line.split("= 0x")
    version = int(version.strip().split(";")[0], 16)

binaries = [file for file in os.listdir(binpath) if file.startswith("AP_FW") and not file.endswith("Pack")]
for file in binaries:
    file = file.strip()
    type_id = -1
    for typeid, typefile in types.items():
        if typefile == file:
            type_id = typeid
            break
    if type_id == -1:
        raise ValueError("We don't recognize filetype <{}>, sorry...".format(file))
    with open(os.path.join(binpath, file), "rb") as binary_file:
        binary = binary_file.read()
        length = len(binary)
        offset = len(output)
        subarr = {
            'type': type_id,
            'version': version,
            'name': file,
            'offset': offset,
            'length': length
        }
        toc.append(subarr)
        output.extend(binary)

jtoc = json.dumps(toc)
tocsize = len(jtoc)
if tocsize > tocmaxsize:
    raise ValueError("TOC is too big! Adjust size and try again.")
output[:len(jtoc)] = jtoc.encode()
with open(os.path.join(binpath, "AP_FW_Pack.bin"), "wb") as output_file:
    output_file.write(bytes(output))

print(toc)
print("All done.")
